### PR TITLE
[Part] Add scaleKnotsToBounds method to BSplineCurve

### PR DESCRIPTION
**Pull request by tomate44**
and change the methods to use dedicated OCCT function

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [x]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by tomate44** [comment]
While thinking back on this, I don't think **setBounds** is a good name choice.
It refers to the bounds() method, that returns parameters and is common to all surfaces.
While these methods are specific to BSplines, and deal with knots.
setKnotsRange(), setKnotsBounds() or setKnotsInterval() would be more accurate, IMHO.
Opinions ?

**Comment by AjinkyaDahale** [comment]
I think setBounds is accurate given the signature. Only the first and last knots are given, and the ones in the middle appear to just be scaled to that range.

However, the topic is different if there already exists a "setBounds" function for other curves (I see that there is a setRange() in some cases) that actually keeps the same parametrization but changes start and end params (effectively changing the shape of the curve). AFAICT a function like this wouldn't make sense for a spline because of it's piecewise nature.

To further complicate the matter, there is a setKnots() function in OCCT that can specify a different sequence of knots. AFAIK that can change the shape of the spline.

So to summarize: setBounds isn't inaccurate, but if a function with the exact same name doing different things exists for other curves, it should be renamed; and, alternative names you suggested can add it's own kind of ambiguity.

If we have to rename I suggest something like "scaleKnotsToBounds()".

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7385">pipeline status</a> for feature branch [PR_7385](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7385). Pipeline [618946924](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/618946924) was triggered at [4129865](https://github.com/FreeCAD/FreeCAD/pull/7385/commits/4129865d40f7187bd2c07de06c3cc9c242e48e3b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by tomate44** [comment]
**scaleKnotsToBounds** is both self-explaining and non-ambiguous.
Thanks a lot.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7385">pipeline status</a> for feature branch [PR_7385](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7385). Pipeline [619281056](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/619281056) was triggered at [8e955ec](https://github.com/FreeCAD/FreeCAD/pull/7385/commits/8e955ec52766dc9f3a5b359ff78df13001c5f829). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7385">pipeline status</a> for feature branch [PR_7385](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7385). Pipeline [619888182](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/619888182) was triggered at [e15ca32](https://github.com/FreeCAD/FreeCAD/pull/7385/commits/e15ca32d079f73dae758f4873988717019bbe6b2). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by AjinkyaDahale** [comment]
@tomate44 is there a reason you chose to make a merge commit rather than a rebase?

**Comment by tomate44** [comment]
@AjinkyaDahale Are you talking about [Merge branch 'master' into BSC_setBounds](https://github.com/FreeCAD/FreeCAD/pull/7385/commits/e15ca32d079f73dae758f4873988717019bbe6b2) ?
On the PR window, Github was saying **'This branch is out-of-date with base branch'** with a **'update branch'** button next to it.
Was it a trap ?

**Comment by AjinkyaDahale** [comment]
@tomate44 not a trap by itself, but if this PR stays for longer there will be another "this branch is out of date", and then the update will add yet another such commit. At least on 2016-17 that was the case.

A cleaner solution is to rebase. This will then rearrange the commits such that it looks as if you made directly on the current master. It can be achieved by clickinh the dropdown next to that "update" button and clicking "rebase and update".

**Comment by tomate44** [comment]
@AjinkyaDahale Thanks for the hint.
I don't know the github web interface very well.
I fixed the problem.

**Comment by AjinkyaDahale** [comment]
> @AjinkyaDahale Thanks for the hint. I don't know the github web interface very well. I fixed the problem.

Yeah I don't use github for this either. Just some tools I'm myself used to (magit within emacs).

**Comment by wwmayer** [comment]
Fixes the test failure: a783d5bbfc
